File: whatsapp-bridge/chat_policy.go
package main

// Conversation allow-list, bridge side.
//
// WHATSAPP_ALLOWED_CHATS (comma-separated JIDs, bare phone numbers, or
// "*@g.us" / "*@s.whatsapp.net" wildcards) restricts which chats the REST API
// will act on. The MCP server enforces the same variable on its tools; the
// bridge repeats the check on every endpoint with a side effect (send, react,
// mark-read, typing) so a bug or a bypass on the MCP side still cannot reach
// a chat you did not enable. Unset means unrestricted.

import (
	"encoding/json"
	"net/http"
	"os"
	"strconv"
	"strings"

	"go.mau.fi/whatsmeow/types"
)

const chatPolicyEnv = "WHATSAPP_ALLOWED_CHATS"

// chatPolicy is the parsed allow-list. Restricted=false means "allow all".
type chatPolicy struct {
	restricted bool
	exact      map[string]struct{}
	servers    map[string]struct{}
}

// normalizeChatEntry canonicalises an allow-list entry or a request target:
// bare number -> phone JID, device suffix dropped, server lower-cased.
func normalizeChatEntry(raw string) string {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return ""
	}
	if !strings.Contains(raw, "@") {
		return raw + "@" + types.DefaultUserServer
	}
	at := strings.LastIndex(raw, "@")
	user, server := raw[:at], strings.ToLower(raw[at+1:])
	if colon := strings.Index(user, ":"); colon >= 0 {
		user = user[:colon]
	}
	return user + "@" + server
}

func parseChatPolicy(raw string) chatPolicy {
	p := chatPolicy{exact: map[string]struct{}{}, servers: map[string]struct{}{}}
	for _, item := range strings.Split(raw, ",") {
		n := normalizeChatEntry(item)
		if n == "" {
			continue
		}
		at := strings.LastIndex(n, "@")
		if n[:at] == "*" {
			p.servers[n[at+1:]] = struct{}{}
		} else {
			p.exact[n] = struct{}{}
		}
	}
	p.restricted = len(p.exact) > 0 || len(p.servers) > 0
	return p
}

func loadChatPolicy() chatPolicy {
	return parseChatPolicy(os.Getenv(chatPolicyEnv))
}

// Allows reports whether the policy permits acting on target (a JID or a bare
// phone number as accepted by the REST API).
func (p chatPolicy) Allows(target string) bool {
	if !p.restricted {
		return true
	}
	n := normalizeChatEntry(target)
	if n == "" {
		return false
	}
	if _, ok := p.exact[n]; ok {
		return true
	}
	_, ok := p.servers[n[strings.LastIndex(n, "@")+1:]]
	return ok
}

// Summary is a one-line description for the startup log.
func (p chatPolicy) Summary() string {
	if !p.restricted {
		return chatPolicyEnv + " unset: all chats allowed"
	}
	n := len(p.exact)
	parts := []string{}
	for s := range p.servers {
		parts = append(parts, "*@"+s)
	}
	desc := strings.Join(parts, ", ")
	if desc != "" {
		desc = " + " + desc
	}
	return chatPolicyEnv + ": restricted to " + strconv.Itoa(n) + " chat(s)" + desc
}

// rejectByChatPolicy writes the 403 the outbound handlers share when target is
// outside the allow-list. Returns true when the request was rejected.
func rejectByChatPolicy(w http.ResponseWriter, policy chatPolicy, target string) bool {
	if policy.Allows(target) {
		return false
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusForbidden)
	_ = json.NewEncoder(w).Encode(map[string]interface{}{
		"success": false,
		"message": "chat " + target + " is not in " + chatPolicyEnv + " (this bridge is restricted to an allow-list of conversations)",
	})
	return true
}

File: whatsapp-mcp-server/chat_policy.py
"""Conversation allow-list (least-privilege access for agents).

``WHATSAPP_ALLOWED_CHATS`` restricts which chats the MCP server will read from
or act on. Unset means everything (today's behaviour). When set, every read
tool filters to the listed chats and every write tool refuses other targets,
so an agent that goes off the rails can only touch what you enabled.

Entries are comma-separated and may be:

- a full JID: ``5511999999999@s.whatsapp.net``, ``120363...@g.us``, ``...@lid``
- a bare phone number (country code, digits only) → ``@s.whatsapp.net``
- a server wildcard: ``*@g.us`` (all groups) or ``*@s.whatsapp.net`` (all DMs)

The bridge enforces the same variable on its outbound endpoints (send, react,
mark-read, typing) as a second line of defence; see whatsapp-bridge/chat_policy.go.
"""

from __future__ import annotations

import os
from collections.abc import Mapping
from dataclasses import dataclass

ENV_VAR = "WHATSAPP_ALLOWED_CHATS"
DEFAULT_USER_SERVER = "s.whatsapp.net"


def normalize_chat_entry(raw: str) -> str:
    """Canonical form for allow-list entries and for JIDs checked against them."""
    value = (raw or "").strip()
    if not value:
        return ""
    if "@" not in value:
        return f"{value}@{DEFAULT_USER_SERVER}"
    user, _, server = value.rpartition("@")
    # Drop a device suffix ("123:4@s.whatsapp.net") so linked-device JIDs match the chat.
    user = user.split(":", 1)[0]
    return f"{user}@{server.lower()}"


@dataclass(frozen=True)
class ChatPolicy:
    exact: frozenset[str]
    servers: frozenset[str]  # from "*@server" wildcards
    restricted: bool

    @classmethod
    def unrestricted(cls) -> ChatPolicy:
        return cls(exact=frozenset(), servers=frozenset(), restricted=False)

    @classmethod
    def from_entries(cls, entries: list[str]) -> ChatPolicy:
        exact: set[str] = set()
        servers: set[str] = set()
        for entry in entries:
            normalized = normalize_chat_entry(entry)
            if not normalized:
                continue
            user, _, server = normalized.rpartition("@")
            if user == "*":
                servers.add(server)
            else:
                exact.add(normalized)
        if not exact and not servers:
            return cls.unrestricted()
        return cls(exact=frozenset(exact), servers=frozenset(servers), restricted=True)

    def allows(self, jid: str | None) -> bool:
        if not self.restricted:
            return True
        normalized = normalize_chat_entry(jid or "")
        if not normalized:
            return False
        if normalized in self.exact:
            return True
        return normalized.rpartition("@")[2] in self.servers

    def sql_clause(self, column: str) -> tuple[str, list[str]]:
        """SQL predicate restricting ``column`` (a chat JID column) to allowed chats.

        Returns ("1=1", []) when unrestricted so callers can always AND it in.
        """
        if not self.restricted:
            return "1=1", []
        parts: list[str] = []
        params: list[str] = []
        if self.exact:
            parts.append(f"{column} IN ({','.join('?' * len(self.exact))})")
            params.extend(sorted(self.exact))
        for server in sorted(self.servers):
            parts.append(f"{column} LIKE ?")
            params.append(f"%@{server}")
        return "(" + " OR ".join(parts) + ")", params

    def denial_message(self, jid: str | None) -> str:
        return f"Chat {jid!r} is not in {ENV_VAR}; this server is restricted to an allow-list of conversations"


def load_chat_policy(env: Mapping[str, str] | None = None) -> ChatPolicy:
    source: Mapping[str, str] = os.environ if env is None else env
    raw = source.get(ENV_VAR, "")
    return ChatPolicy.from_entries([item for item in raw.split(",") if item.strip()])
